web/ts/stream-search.ts

```diff
@@ -0,0 +1,77 @@
+enum SearchDomain {
+    VideoSections = "sections",
+    Streams = "streams",
+    Chats = "chats",
+}
+
+export class StreamSearch {
+    readonly courseId: number;
+
+    timeout: NodeJS.Timeout;
+    loading: boolean;
+
+    results: object[];
+
+    constructor(courseId: number, slug: string) {
+        this.courseId = courseId;
+        this.results = [];
+        this.loading = false;
+    }
+
+    getResults(results: number[][]): object[] {
+        const all = []; // create key value pair from array of array
+
+        results.forEach((result) => {
+            result.forEach((streamId) => {
+                if (!all.includes(streamId)) {
+                    all.push(streamId);
+                }
+            });
+        });
+        return all;
+    }
+
+    reset() {
+        this.results = [];
+        this.loading = false;
+    }
+
+    async getStreamFromResult(streamId): Promise<object> {
+        return await fetch("/api/stream/" + streamId)
+            .then((res) => res.json())
+            .catch((err) => console.log(err))
+            .then((stream) => stream);
+    }
+
+    async performSearch(q: string) {
+        if (q !== "") {
+            clearTimeout(this.timeout);
+
+            this.loading = true;
+            this.timeout = setTimeout(() => {
+                Promise.all([
+                    this.search(q, SearchDomain.VideoSections),
+                    this.search(q, SearchDomain.Streams),
+                    this.search(q, SearchDomain.Chats),
+                ]).then((searchResponses) => {
+                    this.loading = false;
+                    this.results = this.getResults(searchResponses.map((sr) => sr.StreamIds));
+                });
+            }, 250);
```

Comment:
The Timeout is there so that the function isn't called after every `keyup`, but rather the last. 
So for example if I type `h-e-l-l-o`:
After `h`,`e`,`l`,`l` `clearTimeout(timout) `resets the timeout, and after the letter `o` it executes the callback of `setTimeout`.